### ISSUE TITLE
PARADE-CNN

### ISSUE THREAD

**Issue by lironT74** [canjiali/PARADE#29]
hi @canjiali,
Can you please clarify what did you mean in PARADE-CNN?
Window-size 2Xd (pytorch conv2d) transforms every two vectors to a scalar. 
Maybe you meant window size 2X1 (with stride 2X1)?

**Comment by andrewyates** [comment]
hi @lironT74, this should actually be a 1D CNN that takes two passage vectors of length d as input.

**Comment by lironT74** [comment]
@andrewyates got it, thanks.

**Comment by lironT74** [comment]
@andrewyates,
The input size for the CNN head is of shape [batch_size, hidden_size, num_passages] so we have the BERT model hidden size as "in" and "out" channels. I am left with setting kernel size, stride, etc. I tried kernel_size=1 and stride=2 but this feels really wrong. Could you please shed more light?
Thanks

**Comment by andrewyates** [comment]
Hi @lironT74, I'm not certain because I don't know pytorch convs offhand, but I think you want kernel_size=stride=2 with the shape [batch_size, num_passages, hidden_size]. Code from tensorflow is below if that helps. Essentially the layer should take two 768-dim passage CLS representations as input and output one new 768-dim representation.

```
    def aggregate_using_convstack(self, expanded_cls):
        remaining_passages = self.num_passages
        while remaining_passages > 1:
            assert remaining_passages % 2 == 0, "num_passages must be a power of 2"
            # from earlier:
            # self.conv = tf.keras.layers.Conv1D(filters=self.bert_config.hidden_size, kernel_size=2, strides=2)
            # expanded_cls = tf.reshape(cls, [batch_size, self.num_passages, self.bert_config.hidden_size])
            expanded_cls = self.conv(expanded_cls)
            remaining_passages /= 2
            
        return expanded_cls
```

**Comment by andrewyates** [comment]
Hi @lironT74, I'm not certain because I don't know pytorch convs offhand, but I think you want kernel_size=stride=2 with the shape [batch_size, num_passages, hidden_size]. The layer should take two 768-dim passage CLS representations as input and output one new 768-dim representation. In TF this is `tf.keras.layers.Conv1D(filters=self.bert_config.hidden_size, kernel_size=2, strides=2)`

**Comment by lironT74** [comment]
@andrewyates Thanks. I'll try that.
